### ISSUE TITLE
Storage for powerups

### ISSUE THREAD

**Issue by maxteufel** [SuperTux/supertux#473]
Store powerups instead of replacing the current powerup.

How to do this in the code (notes for myself):
- store stuff in some place
- menu for accessing in pause menu
- details to be added

**Comment by J-128** [comment]
Sounds like a great idea to me!

**Comment by maxteufel** [comment]
This could be used to realize storage of collectibles planned for the M3 storyline.

Also, add a scripting API.

**Comment by J-128** [comment]
So it could! Great thinking!

**Comment by Karkus476** [comment]
@maxteufel 
Remember that this should also be:
 [ ] Per-worldmap
 [ ] Per-profile
 [ ] Saved when you quit

**Comment by Karkus476** [comment]
Wait, if your fire and you pick up ice you stay fire? Or you become ice but store fire?

**Comment by christ2go** [comment]
Storage space should be limited, and maybe only level-local, else this could give too much power to one player. And some levels are designed to only use one (e.g.) fire flower, because with the ice flower they would be too easy, I'd say it should be only level-local.

**Comment by maxteufel** [comment]
@christ2go This is a good point and will be definitely considered when implementing this feature. However, this is also intended to be used for storing other objects (e.g. a better API for e.g. keys in the Forest World). This requires per-world storage.

**Comment by Karkus476** [comment]
@christ2go You can take whatever items you want into any level, if they exist in the same worldmap. It should be worldmap specific.

**Comment by christ2go** [comment]
I implemented this feature (see my branch storepowerup), and while playing some levels noticed, that this feature is giving the player far too much power (i played the last 6 levels, whenever I had to shrink loaded a saved fireflower, and in the end I still had 8 fireflowers left). If this feature should be added, the number of powerups in all levels would need to be adjusted.

**Comment by maxteufel** [comment]
@christ2go Can you test it with "Power-up menu (Choose what you want)" only before levels, and only keep one next powerup in the queue available for a single level? (What Mario games do)

**Comment by christ2go** [comment]
@maxteufel Do you mean like the store in super marico chronicles? (One powerup always replaced?)

**Comment by maxteufel** [comment]
- Keep a list of powerups and numbers available (could be bought in a shop or awarded for winning minigames or so in the future)
- Current level:
  - no powerup + powerup => you get powerup
  - powerup + same powerup => powerup increased
  - powerup + different powerup => keep powerup, store different powerup
    - different powerup can be used on request in current level
    - if not used until end of level, awarded to global powerup store

**Comment by christ2go** [comment]
@maxteufel Is the select powerup dialogue in front of every level still part of this?

**Comment by maxteufel** [comment]
@christ2go Yes, but only provide that global storage before starting the level, not when in it. Also, that shouldn't happen automatically only when requested (add a menu point in the worldmap-escape-menu)

**Comment by Karkus476** [comment]
@maxteufel powerup +same_powerup when powerup is earth or air needs to be considered separately

**Comment by maxteufel** [comment]
@Karkus476 as for the earth powerup, having more of it extends the time the stone feature can be used.

**Comment by Karkus476** [comment]
@maxteufel I see. We need a way to teach the player these things (Including more fire powerups means more fireballs.)

**Comment by maxteufel** [comment]
More introductory levels like "Welcome to Antarctica" and "A Fork in the Road"? Some of them should be in forest, I suppose.

**Comment by christ2go** [comment]
When implementing this feature I found a few points, which still might need considerations. 
- Should Tux really not be allowed to store the same powerup that he currently has?
- Currently powerup loading is implemented as a switch powerup mechanism, the question is, should Tux also be able to "switch" when no current powerup is stored (so he can become small for a minor part in the level and later become big again)
- I thought that powerup loading before a level might be done in the level loading screen, by using the left and right controls Tux can switch between all the powerups he can load.

If the two upper points shouldn't be implemented, this might have the advantage, that the storage isn't usable until introduced and wouldn't make any existing level easier.

**Comment by maxteufel** [comment]
> - Should Tux really not be allowed to store the same powerup that he currently has?

Then extending the powerup's abilities based on how many of it you have won't make sense anymore (e.g. right now with a higher number of fire/iceflowers, you can shoot more in the same time; more earthflowers = longer time stone feature can be used; etc.).

**Comment by christ2go** [comment]
@maxteufel  Thanks, that would be clear. What about switching anytime and loading screen?

**Comment by maxteufel** [comment]
@christ2go Sounds like a good idea, if the UI/UX is implemented properly. We don't want non-understandable hidden switches to do that. It should be clearly visible that you can add powerups and how you can do that and what consequences it will have.

**Comment by Karkus476** [comment]
This needs to be scriptable (make use of your variable storage file, *.sss, so that load and store could be used to read the items the user has?) and general, so that anything can be stored (keys, other map items or secret items from levels). Pressing action on the Worldmap should bring up the UI, which will need to show icons of the items you have stored, in a grid formation. The items can create stacks to save space, if this is needed (so for example you could store 3 of the same item in the same spot on the grid). What you've done already is great, but it will take a nice UI to bring it together

**Comment by maxteufel** [comment]
> (make use of your variable storage file, *.sss, so that load and store could be used to read the items the user has?)

No. Store the main game powerups it in the `*.stsg` as that is used for core progress. 'Custom powerups' or 'custom collectibles' should be stored in the `*.sss`.

Indeed, this needs a good UI, as @Karkus476 notes.

**Comment by maxteufel** [comment]
As for custom variants, a sprite (for UI) and name (for scripting) should obviously be stored.

**Comment by brmbrmcar** [comment]
If there were different powerups, I think that the one collected last should be used, but if the older powerup was collected again it would go back, being more powerful.

**Comment by christ2go** [comment]
@brmbrmcar There are different powerups, or what do you mean?

**Comment by brmbrmcar** [comment]
So, say you had fire, then if you got ice you would be ice. But if you then got fire, you will have double power fire or whatever because you had fire once before.

**Comment by christ2go** [comment]
Yeah, that's right you could theoretically still get the same powerup stored ... But only in levels were different powerups are in place, so it wouldn't make existing levels more easy to play. But honestly it's kind of hard to explain to the player that if he has fire *collected* and he collects fire again it will just be used to make his newly collected fire hat more powerful.

**Comment by brmbrmcar** [comment]
It is hard to explain. But what could happen is some kind of change to the hat every time something is more powerful. Then it would just seem logical.

**Comment by LizzyBach** [comment]
> Wait, if you're fire and you pick up ice you stay fire? Or you become ice but store fire?

I'm really confused. What do you mean?

**Comment by LizzyBach** [comment]
> > * Should Tux really not be allowed to store the same powerup that he currently has?
> 
> Then extending the powerup's abilities based on how many of it you have won't make sense anymore (e.g. right now with a higher number of fire/iceflowers, you can shoot more in the same time; more earthflowers = longer time stone feature can be used; etc.).

Maybe, it'll require some more optimizations, I think?

**Comment by Guih48** [comment]
I think powerup stackind mechanics(fireball/iceball count) and powerup storage/switching is could be a good combination if the storage is replaced by a switch, that switches between powerups, because that mechanic: 

If you have 8 fireballs, and you gain an ice flower the fireball count is saved and if you gain a fire flawer the count is starts from 8 (+1 because you gained a new flower) and the iceball count is also saved (this only works only if we didn't hurt!).

And this switch could switch between saved counts (at worldmap) without we gain a new powerup.

**Comment by Paulo-Tux** [comment]
In my opinion, I think we should click a button on the keyboard to throw a fireball and click another button on the keyboard to throw an iceball so that Tux has the power of fire and ice.

**Comment by bruhmoent** [comment]
This has been addressed by the `Item Pocket`, closing.
